site/app/libraries/database/DatabaseQueries.php

```diff
@@ -1407,6 +1407,102 @@ public function getLateDayInformation($user_id) {
         return $this->course_db->rows();
     }
 
+    public function getLateDayCache() {
+        $query = "SELECT * FROM 
+                    late_day_cache AS ldc
+                    LEFT JOIN gradeable g ON g.g_id=ldc.g_id
+                  ORDER BY late_day_date NULLS LAST, g.g_id NULLS FIRST";
+        $this->course_db->query($query);
+        $return = [];
+
+        // key: user, value: array of all cache available
+        foreach ($this->course_db->rows() as $row) {
+            $user_id = $row['user_id'] ?? $row['team_id'];
+            // title = g_title or event date
+            $title = $row['g_title'] !== null ? $row['g_title'] : explode(" ", $row['late_day_date'])[0];
+            $return[$user_id][$title] = $row;
+        }
+        return $return;
+    }
+
+    public function getLateDayCacheForUser($user_id) {
```

Comment:
It would be good to add type hints to the parameters and return value for this function

Reply:
Perhaps Matt can correct me if I'm wrong here but in my opinion, it would be better to use [this](https://www.php.net/manual/en/language.types.declarations.php) type of type hinting.  Using the above method instead of comments, any parameters that are not of the proper type will result in PHP errors in the development environment which helps to catch bugs earlier in my experience.

Reply:
I see what you mean. In that case do you think I should remove the docstring or keep it?

Reply:
For simple things where the docstring duplicates the information provided by the type hint, I would suggest removing it because it simply clutters the code.  For things where you have specified the return type in more detail, I would argue that it would be good to keep it so that someone can quickly reference it in the future should they wander into this part of the code and wonder what that function returns.  

Matt may have a better suggestion here though.

Reply:
I see some other functions/constructors with the duplicate type hinting, so I will keep it for now. Thanks!